Pigeon multiple arity generator

gaaclarke opened the pull request:
fixes https://github.com/flutter/flutter/issues/86971
## Pre-launch Checklist

- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated `CHANGELOG.md` to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

gaaclarke commented:
@stuartmorgan This is ready to review.  Most of this code was written before we resolved the last PR so there might be some residual issues I didn't update after the fact.

The one thing we didn't resolve from our previous conversation is the naming structure for objc selectors from dart function declarations.  There was talk about being able to specify the argument labels which is a good idea but I don't think it should be required, so we still need to come up with a default transform.

Right now it's:
```dart
int add(int x, int y);
int inc(int x);
```

```objc
-(void)add:(NSNumber*)x y:(NSNumber*)y completion:((^)(NSNumber* result, FlutterError* error)completion;
-(void)inc:(NSNumber*)x completion:((^)(NSNumber* result, FlutterError* error)completion;
```

There are some cases where it is a bit weird but it is pretty simple and predictable to me.

stuartmorgan commented:
> There are some cases where it is a bit weird but it is pretty simple and predictable to me.

It's simple, but it almost always violates one of the core [ObjC naming rules](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html): "Make the word before the argument describe the argument". Including the first argument name as well is just as simple, and will be slightly less wrong. In both of your simple examples, the result is better, for instance:
- `addX:Y:`—should actually be something like `addInt:toInt:` or `addValue:toValue:`, but `addX:toY:` would be livable, and `addX:Y:` is closer to that. And `add(int value, int value2)` would then give something approaching the right name.
- `incX:`—should be `incrementInt:` or `incrementValue:`, and `increment(int value)` would be a completely reasonable Dart declaration that would then give the right ObjC name

> There was talk about being able to specify the argument labels which is a good idea but I don't think it should be required

Per discussion, I'm fine with it not being required for Pigeon users, but I absolutely think it's required that Pigeon support it. Most of the auto-generated names are going to be deeply un-idiomatic ObjC names, and we shouldn't be doing code generation whose only possible output is bad code that the developer can't tune to make into a reasonable, idiomatic API surface to write their native code again.

stuartmorgan commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -107,8 +108,28 @@ String _makeGenericCastCall(TypeDeclaration type, String nullTag) {
+) {
+  return func.arguments.isEmpty
+      ? ''
+      : map2(
```
It looks like_spaceJoin is the only thing you ever use `map2` for. You could avoid introducing a non-standard collection operation for readers to go look up by just doing a `zip`+`map` instead. You can then also trivially inline _spaceJoin instead of defining it. So it becomes:

```
  zip(
    func.arguments
        .map((NamedType e) => _addGenericTypes(e.type, nullTag)),
    argNames,
  )
  .map((Tuple t) => 't.first t.second')
  .join(', ');
```

That's not only less code overall, but completely understandable locally (assuming someone knows `zip`, but that's a pretty standard method in collection utilities) instead of having to look at `map2` and `_spaceJoin`.

gaaclarke replied:
Allowing `map` to take in multiple arguments is pretty standard, both scheme and clojure allow it (I think some others too).  In common-lisp it's called `mapcar` which we shouldn't use haha. 

While writing this code I realized there was some overlap in use between zip and map2 (since zip is a special case of map2).  The reason I kept map2 is because of performance, avoiding allocations/deallocations, and it's slightly easier to work with since you don't have to consider what `.first` and `.second` actually mean (That wouldn't be as big of a deal if Dart could deconstruct tuples like F#).

stuartmorgan replied:
> Allowing map to take in multiple arguments is pretty standard, both scheme and clojure allow it

FYI, I'm pretty sure you are dramatically overestimating the percentage of developers who have any meaningful exposure to functional programming languages.

I'm not saying we can't adopt functional approaches where it's useful but "it's a standard thing someone reading this code should know" *definitely* does not follow from "Scheme and Lisp allow it".

gaaclarke replied:
Haha fair enough, my examples will always be biased by my experience.  The question of if it's a standard function is a bit of a red herring to me.  The important questions are the ones we ask of any function: Is it useful?  Is it understandable?  Is it clearly documented?

gaaclarke replied:
FWIW python does this as well: https://docs.python.org/3/library/functions.html#map

stuartmorgan replied:
> it's slightly easier to work with since you don't have to consider what `.first` and `.second` actually mean

I'm not sure I'm following this argument. Are you saying that going and looking up `_spaceJoin` and figuring out what `x` and `y` actually mean is easier than looking at the tuple created in the previous statement to see which argument is `first` and which is `second`?

`x` and `y` are both less meaningful, and further removed in the code from the context, so I'm not sure how that would be easier.

gaaclarke replied:
Yep, `x` and `y` are only needed to understand the abstraction of `_spaceJoin`.  After I understand what `_spaceJoin` does they are not required for understanding `_getMethodArgumentsSignature`.  Whereas with a zip call I have to reference a different point in code to know what `.first` and `.second` refer to.  I'd say map2 is always preferable to zip in Dart.

stuartmorgan commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -107,8 +108,28 @@ String _makeGenericCastCall(TypeDeclaration type, String nullTag) {
+              .map((NamedType e) => _addGenericTypes(e.type, nullTag)),
+          argNames,
+          _spaceJoin,
+        ).join(', ');
```
While writing the above, I realized that I'm confused about this method at a higher level: I was wondering why the argument names were a separate iterator in the first place, and pieced together the various calls, and unless I'm mis-reading you are:
- Starting with `func.arguments`
- Having the caller pull out just the names via _getMethodArgumentNames to create `argNames`
- Having this pull out just the types via line 127
- Zipping them back into one piece at the end.

Why isn't this a function that just takes `func.arguments` and `nullTag` as parameters, and constructing the string directly from func.arguments in a simple iterator? Doing the two halves separately, and at different layers of the code, seems like it's just making it unnecessarily hard to follow what's happening.

gaaclarke replied:
> I was wondering why the argument names were a separate iterator in the first place, and pieced together the various calls

If I'm understanding the question correctly, the reason that `argNames` is a parameter is that it guarantees that the same argument names are used here that are used elsewhere in the code.  See line 164:
```dart
        final Iterable<String> argNames = _getMethodArgumentNames(func);
        sendArgument = '<Object>[${argNames.join(', ')}]';
        argSignature = _getMethodArgumentsSignature(func, argNames, nullTag);
```

`argNames` is used twice for 2 different calculations.  Does that answer your question?  Technically `_getMethodArgumentsSignature` could call `_getMethodArgumentNames` but that would be a duplication of that logic that could create problems in the future.

stuartmorgan replied:
We should discuss this offline tomorrow.

(To preview at a high level: it is not clear to me what future problem that could create that is worse than allowing someone to call `_getMethodArgumentsSignature` where `_argNames` is anything other than exactly `_getMethodArgumentNames(func)`, which your current structure makes trivial. E.g.,, nothing would stop someone from calling `_getMethodArgumentsSignature(someFunc, _getMethodArgumentNames(aCompletelyDifferentFuncWithTheSameArgCount), nullTag)` with incredibly bad results.)

gaaclarke replied:
Done.  I talked with stuart offline.  His biggest concern was relying on the 2 input data structures being passed into `_getMethodArgumentsSignature` having to mirror each other.  My concern was having different places in the code that aren't next to each other that have to match.  By making `_getMethodArgumentsSignature` a higher order function we could solve both problems.  Now it looks like this:
```dart
        final Iterable<String> argNames =
            indexMap(func.arguments, _getArgumentName);
        sendArgument = '<Object>[${argNames.join(', ')}]';
        argSignature =
            _getMethodArgumentsSignature(func, _getArgumentName, nullTag);
```
If I wanted to be extra secure I could pull _getArgumentName into a local variable and share it, but this is fine.

stuartmorgan commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -252,15 +275,27 @@ void _writeFlutterApi(
+                    func.arguments, (int count, NamedType arg) => 'arg$count');
+                final Iterable<String> argTypes = func.arguments
+                    .map((NamedType e) => _addGenericTypes(e.type, nullTag));
+                enumerate(zip(argTypes, argNames),
```
Same high-level question here. Why start with func.arguments, separate it into two iterators that you have to know go together by index, then re-zip them, rather than just doing a simple `enumerate` with string conversion? Isn't this all just:

```
enumerate(func.arguments, (int count, NamedType arg) {
  final String argType = _addGenericTypes(e.type, nullTag);
  final String argName = 'arg$count';
  [writeln code here, unchanged]
}
```

Slightly less code, no `zip`, and less mental state that the reader needs to keep (the two temp iterators).

stuartmorgan replied:
(I expect these comments will apply to the other languages as well; whatever resolution we get to here should be applied to them as well, but I won't leave comments there for now to avoid fracturing discussion).

gaaclarke replied:
Same thing, `argNames` is used it 2 separate calculations so I don't want to duplicate the logic to calculate them and instead rely on basing everything off of one calculation stored in the `argNames` variable.

stuartmorgan replied:
Did you want to leave this one unchanged, rather than resolving it the same way as in the other case using `_getSafeArgumentName`?

I'm okay with it if you feel strongly about it, since in this case the two usages are entirely local, but I do think even here it would be a bit more understandable with duplicate delegation to a helper instead of the split-enumerate-zip-re-enumerate flow.

gaaclarke replied:
Sure, done.

stuartmorgan commented on packages/pigeon/mock_handler_tester/test/message.dart:
```diff
@@ -113,17 +143,16 @@ class NestedApi {
       );
     } else {
-      return SearchReply.decode(replyMap['result']!);
+      // noop
```
What's the purpose of generating a no-op else branch? There's no explanation of why it's a no-op, so it doesn't seem as an unfamiliar reader to be adding any useful information.

gaaclarke replied:
Typically that is where the return statement goes.  Since this is a void returning function there is none.  I don't want to switch the structure of the function depending on wether it returns void or not so I'll have that else branch.  The `noop` was added to make it not look like a mistake, that it wasn't intentional.

stuartmorgan replied:
> I don't want to switch the structure of the function depending on wether it returns void or not so I'll have that else branch.

Why not? I found the generated code more confusing than if this weren't here at all. It seems like you're optimizing for ease of implementation at the cost of making the generated code (which far more people will see) more confusing.

But if you really want to keep parallel structure for some reason, why not `return;` here instead? Then the structure is more similar.

gaaclarke replied:
Done. `return;` sounds fine with me.  I try to make the generated code readable but sometimes I'd say simplicity of the generator trumps output style.

stuartmorgan commented on packages/pigeon/mock_handler_tester/test/message.dart:
```diff
@@ -164,7 +235,63 @@ class Api {
+          assert(message != null,
+              'Argument for dev.flutter.pigeon.FlutterSearchApi.search was null.');
+          final List<Object?> args = (message as List<Object?>?)!;
+          final SearchRequest? arg0 = args[0] as SearchRequest?;
```
Why are these arg0, arg1, etc. rather than using the actual argument name? I know it's generated code, but it seems like it would be simple to use the name (prefixed to avoid collisions with variables you are declaring) so that someone who does read this code for some reason (say, while debugging issues that turn out to be in the generated code) would have an easier time.

gaaclarke replied:
I did it on purpose to avoid any possibility for conflicting name types.  You do make a good point about being easier to debug though.

gaaclarke replied:
Done.

stuartmorgan commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -252,15 +275,27 @@ void _writeFlutterApi(
+                  indent.writeln(
+                      'final $argType$nullTag $argName = $argsArray[$count] as $argType$nullTag;');
+                  indent.writeln(
+                      'assert($argName != null, \'Argument for $channelName was null. Expected $argType.\');');
```
How about `Expected a value of type "argType"`. I found that the generated error messages read oddly as-is, since `null` is a value but this is a type.

gaaclarke replied:
Reworded to "Argument for com.foo.bar was null, expected non-null int."

FWIW this shouldn't be able to be triggered by our users, this should only happen as a result of an error inside of the pigeon code (or someone abusing the generated code).

gaaclarke commented:
@stuartmorgan give this issue a look.  I think this is what we talked about previously: https://github.com/flutter/flutter/issues/88445

stuartmorgan commented on packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/MultipleArityTest.m:
```diff
@@ -0,0 +1,36 @@
+      [[MultipleArityFlutterApi alloc] initWithBinaryMessenger:binaryMessenger];
+  XCTestExpectation *expectation = [self expectationWithDescription:@"subtraction"];
+  [api subtract:@(30)
+               y:@(10)
```
I know it's just a test, but maybe the example multiple-arity function could be something that's not intrinsically a poorly-named function? The ObjC generated really highlights it, but even in other languages `subtract(int x, int y)` is a terrible function because there's nothing anywhere in the declaration that indicates the—critically important—order. Just something that's not order-dependent (e.g., `multiply`) would be a big improvement.

gaaclarke replied:
I purposefully chose a function that wasn't transitive to make sure that the order was being asserted.  I originally wrote this as `add` but it didn't test the code as well as `subtract`.

stuartmorgan replied:
Maybe it should be something whose arguments are different types? That would not only test order, but would make order problems a compile error rather than a runtime error. E.g., `getDisplayString(int value, String unit)` (`=> '$value$unit'`)

gaaclarke replied:
Runtime versus compilation error isn't a problem for unit tests since a runtime error happens a fraction of a second later.  It's not trivial to change this example, 30-40 minutes.  There are precedents for functions like this, `atan2` and `pow`.  Having a poor objc output will be addressed in the custom selector PR.

stuartmorgan replied:
To clarify: my first comment is the thing I think is the problem here (the terrible function name). The second was a suggestion for a specific fix that in addition to resolving that, would both address your response (that the function needs to assert order) and have secondary benefits (compile-time error).

> There are precedents for functions like this, atan2 and pow

`atan2` does indeed have this problem. `pow` does not because `num pow(num x, num exponent)` has unambiguous argument names—nobody could read that and not know which of the two arguments is the exponent. I wouldn't have an issue with this example if it were `pow(int x, int exponent)`.

But if you don't care that the example API in your tests is a naming anti-pattern, okay.

stuartmorgan commented on packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/EchoMessenger.h:
```diff
@@ -8,6 +8,8 @@
 NS_ASSUME_NONNULL_BEGIN
 
 @interface EchoBinaryMessenger : NSObject<FlutterBinaryMessenger>
```
Please comment all classes, per style guide.

gaaclarke replied:
Done

stuartmorgan commented on packages/pigeon/test/java_generator_test.dart:
```diff
@@ -764,4 +764,68 @@ void main() {
+    expect(
+        code,
+        contains(
+            'public void add(Long arg0, Long arg1, Reply<Long> callback)'));
```
Why aren't these real argument names?

gaaclarke replied:
I never implemented it for java.  I can do that now, it should be a small change I think.

gaaclarke replied:
Oh wait, I was thinking this was for the interface.  I already do this for the interface.

The reason I didn't allow user specified names here is because I didn't want them to collide with other names in that function.  If they named the arguments `output` or `channel` for example it would break the code.

stuartmorgan replied:
As mentioned in the other comment, you could prefix to avoid that problem. E.g., `arg_$argName`.

gaaclarke replied:
Done.  When running into this problem in non-hygienic macros you typically use something like `gensym` to make sure you get a unique symbol for the local variable.  I made a simple `_generateSymbol` function that just prefixes them with `pigeon_`.

stuartmorgan replied:
I'm still seeing a bunch of places in the generated code (both ObjC and Java) where things are called arg0, arg1, etc. Can't they all be prefixed versions of the names?

gaaclarke replied:
Yea, I only implemented it for `@FlutterApi()` + Java.  It looks like it still happens for `@HostApi()` + Dart and some of the files just haven't been regenerated.  I should be able to duplicate what I did here across the other languages, shouldn't be a big deal.  We'll say this is done when 'arg0' doesn't show up in the diff anywheres.

gaaclarke replied:
I went back through and audited everything.  All user facing variables have user supplied names.  The leftover `arg0` arguments are in generated code that don't face the user.  You mention in another comment that those could be easier to debug if they had names so I'll make that change.

gaaclarke commented on packages/pigeon/test/functional_test.dart:
```diff
@@ -0,0 +1,70 @@
+        expect(value, 'c');
+        saw |= 0x4;
+      }
+    }).toList();
```
Done, I switch the test to check the output.

gaaclarke commented on packages/pigeon/test/functional_test.dart:
```diff
@@ -0,0 +1,70 @@
+    expect(result[0].second, 'a');
+    expect(result[1].second, 'b');
+    expect(result[2].second, 'c');
+  });
```
Done.

gaaclarke commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -107,8 +108,28 @@ String _makeGenericCastCall(TypeDeclaration type, String nullTag) {
 
-String _getArgumentName(NamedType field) =>
-    field.name.isEmpty ? 'arg' : field.name;
+String _getArgumentName(int count, NamedType field) =>
```
In my mind they are both required data for the calculation.  The only reason I'll prefer one order over another is if I'm trying to mimic object oriented programming where it isn't available (like C) or if it's a higher order function I might put the function parameter at the end to make it a bit cleaner when passing in anonymous closures.  Otherwise it's just not a thought.

stuartmorgan replied:
If you don't care, then can we swap it? I consider it pretty standard practice to have `getSomething(primarySource, fallback)` rather than `getSomething(fallback, primarySource)`.

gaaclarke replied:
I just went to do it and there is another reason that they are in this order that I didn't see, it's the order required by `indexMap`.  If i change the order I'll have to map a lambda function just to swizzle the order of parameters.

gaaclarke commented on packages/pigeon/lib/functional.dart:
```diff
@@ -0,0 +1,53 @@
+  int index = 0;
+  for (final T value in iterable) {
+    yield func(index, value);
+    index += 1;
```
Done.  I thought people prefer `+=` over `++`, I switched it.  I do prefer it being its own statement so people don't even have to think about order of operations.

gaaclarke commented:
@stuartmorgan I think there are a few things outstanding that we'll talk about tomorrow.  Can you hit the "resolve conversation" for your comments that are resolved?  I don't what to presume they are resolved until you get a chance to look.

gaaclarke commented:
@stuartmorgan Okay, I implemented the changes we talked about PTAL.  There are still some references to `arg0` but that is because of tests that don't have named arguments.  I think it's technically impossible to create them from the front end parser but I support it in the backend for now.

I pulled out the other issue we talked about: https://github.com/flutter/flutter/issues/88545

gaaclarke commented:
I was just playing with this and it looks like this isn't quite right yet, having allowed the user defined named opened another possibility for collisions:
```dart
class Api {
  /// Constructor for [Api].  The [binaryMessenger] named argument is
  /// available for dependency injection.  If it is left null, the default
  /// BinaryMessenger will be used which routes to the host platform.
  Api({BinaryMessenger? binaryMessenger}) : _binaryMessenger = binaryMessenger;

  final BinaryMessenger? _binaryMessenger;

  static const MessageCodec<Object?> codec = _ApiCodec();

  Future<int> add(int x, int y) async {
    final BasicMessageChannel<Object?> channel = BasicMessageChannel<Object?>(
        'dev.flutter.pigeon.Api.add', codec, binaryMessenger: _binaryMessenger);
    final Map<Object?, Object?>? replyMap =
        await channel.send(<Object>[x, y]) as Map<Object?, Object?>?;
    if (replyMap == null) {
      throw PlatformException(
        code: 'channel-error',
        message: 'Unable to establish connection on channel.',
        details: null,
      );
    } else if (replyMap['error'] != null) {
      final Map<Object?, Object?> error = (replyMap['error'] as Map<Object?, Object?>?)!;
      throw PlatformException(
        code: (error['code'] as String?)!,
        message: error['message'] as String?,
        details: error['details'],
      );
    } else {
      return (replyMap['result'] as int?)!;
    }
  }
  ...
```

I think I'm going to go back to the idea of prefixing those with `arg_` instead of trying to make sure all locally available functions are prefixed with `pigeon_`.

gaaclarke commented:
Okay done.  There are no unmodified argument names in functions that have implementations.  In interfaces and headers I leave them as-is.  The only other change I made from what we discussed is I used the `fooArg` naming scheme for java, not `arg_foo` since the Java linter complains about those names.

stuartmorgan commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -107,8 +108,26 @@ String _makeGenericCastCall(TypeDeclaration type, String nullTag) {
 
-String _getArgumentName(NamedType field) =>
-    field.name.isEmpty ? 'arg' : field.name;
+/// Returns an argument name that can be used in a context where it is unlikely to collide.
```
Nit: don't you mean "where it has the potential to collide"? The non-`Safe` version would be the one for contexts where you don't need to worry about collisions.

gaaclarke replied:
Done.

stuartmorgan commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -252,15 +275,29 @@ void _writeFlutterApi(
+                final Iterable<String> argNames = indexMap(
+                    func.arguments,
+                    (int count, NamedType arg) =>
+                        arg.name.isEmpty ? 'arg$count' : 'arg_${arg.name}');
```
Unless I'm missing something, this lambda is a direct duplicate of `_getSafeArgumentName`.

gaaclarke replied:
Done.

stuartmorgan commented on packages/pigeon/lib/java_generator.dart:
```diff
@@ -158,16 +167,21 @@ static MessageCodec<Object> getCodec() {
+                  enumerate(method.arguments, (int index, NamedType arg) {
+                    final String argType = _javaTypeForDartType(arg.type);
+                    final String argName =
+                        arg.name.isEmpty ? 'arg$index' : '${arg.name}Arg';
```
How about declaring this as a function with a comment explaining it, as you did in the Dart code? I think it would non-obvious to a reader who didn't just have the discussions we've just had about collisions why this isn't just calling `_getArgumentName`, whereas the commented function makes it very clear.

gaaclarke replied:
Done.

stuartmorgan commented on packages/pigeon/lib/java_generator.dart:
```diff
@@ -244,27 +263,35 @@ static MessageCodec<Object> getCodec() {
+        final Iterable<String> argTypes =
+            func.arguments.map((NamedType e) => _javaTypeForDartType(e.type));
+        final Iterable<String> argNames = indexMap(func.arguments,
+            (int index, NamedType arg) => _getArgumentName(index, arg) + 'Arg');
```
This could also then share that function instead of duplicating the appending of `Arg`.

gaaclarke replied:
Done.

stuartmorgan commented on packages/pigeon/lib/objc_generator.dart:
```diff
@@ -232,47 +234,67 @@ NSObject<FlutterMessageCodec>* ${_getCodecGetterName(options.prefix, api.name)}(
+  required String returnType,
+  required String lastArgLabel,
+  required String lastArg,
+  Iterable<String>? argNames,
```
This is the same construction that we agreed not to use in the Dart code; it should be changed here as well.

gaaclarke replied:
Done.

stuartmorgan reviewed (changes requested):
A few comments on the changes, but the one you changed per discussion looks much better, thanks.

There are also a few unaddressed comments from earlier (e.g., [the default ObjC signature generation](https://github.com/flutter/packages/pull/426#issuecomment-901129583))

gaaclarke commented:
@stuartmorgan I'm going to address the custom objc selectors in a separate PR.  That's why I spun out https://github.com/flutter/flutter/issues/88445.  This one just has the logic for the default behavior.

stuartmorgan commented:
> @stuartmorgan I'm going to address the custom objc selectors in a separate PR. That's why I spun out [flutter/flutter#88445](https://github.com/flutter/flutter/issues/88445). This one just has the logic for the default behavior.

I'm aware of that, but the majority of [the comment I was referring to](https://github.com/flutter/packages/pull/426#issuecomment-901129583) is about the default behavior, and it hasn't been addressed.

gaaclarke commented:
@stuartmorgan and I talked offline and agreed to accepting the breaking change and doing:
```objc
-(NSNumber*)addX:(NSNumber*)x y:(NSNumber*)y;
-(NSNumber*)printValue:(Input*)value;
```

That is now implemented.

**However** in the course of migrating the tests I've changed my mind about this breaking change.  The reason is that the compiler for Objective-C is a pretty permissive about certain mistakes which result in runtime errors instead of compilation errors.  In this instance, not implementing a protocol you say you implement by default is a warning instead of an error.  When I accepted the breaking change I assumed we'd only introduce compilation errors.

If you look at this example: https://github.com/flutter/packages/pull/426/files#diff-0d354441e98e282ffa2dd5a9836936b4d32be5d13ba8f97635ec3a4e2e69e794L61  When I implemented the breaking change, most of the tests created compilation errors which was wasn't bad to fix.  But tests based on that mock object started experiencing runtime errors which I had to debug (unrecognized selector exceptions).

I suggest we keep the refactoring around I did to implement this but switch back to the original naming scheme.

stuartmorgan commented:
> **However** in the course of migrating the tests I've changed my mind about this breaking change. The reason is that the compiler for Objective-C is a pretty permissive about certain mistakes which result in runtime errors instead of compilation errors. In this instance, not implementing a protocol you say you implement by default is a warning instead of an error. When I accepted the breaking change I assumed we'd only introduce compilation errors.

Is the concern here that people will run the new version of pigeon, not pay attention to the CHANGELOG, ignore compiler warnings, not test their plugin, and thus ship code with runtime errors? I know breaking changes aren't zero-cost, but if we are unwilling to make them because a small number of users of an early-stage, pre-1.0 project might be burned by them if they have very lax engineering practices, we're setting the bar way too high.

> I suggest we keep the refactoring I did to implement this but switch back to the original naming scheme.

The fundamental tradeoff here is still: do we make Pigeon's default behavior worse forever (you're certainly not going to more inclined to make this breaking change later, when there are presumably more users), for all future users, or do we maybe break a very small number of people now. I'm still firmly in the second camp, and that's consistent with Flutter's philosophy on breaking changes that was informed by user feedback.

If you really think that the fallout of the breakage is that severe, you could implement logic in the generator that instead of producing code like:
```
  [foo addValue:theValue];
```

produces:
```
  if (![foo respondsToSelector:@selector(addValue:) && [foo respondsToSelector:@selector(add:)) {
    NSLog(@"WARNING: You need to update your code, see <insert helpful URL here>");
    [foo add:theValue];
  } else {
    [foo addValue:theValue];
  }
```

I don't personally think that's worth the effort for the extremely small payoff, but if that's the only way you're comfortable making the change to make the default a) more correct, and b) more consistent, then we should do that, not scrap the change.

gaaclarke commented:
@stuartmorgan Not just that they will ship runtime errors, which is a possibility, but the cost to upgrade.

Here's what we can do, the *Setup functions can assert that the passed in class actually conforms to the protocol.  That way they'll get an apropos error message at the earliest point in time, instead of relying on them to call the right message to discover the error.  Who knows, that message might only fire once in a blue moon.

stuartmorgan commented:
> Here's what we can do, the *Setup functions can assert that the passed in class actually conforms to the protocol. That way they'll get an apropos error message at the earliest point in time, instead of relying on them to call the right message to discover the error.

That sounds good to me.

gaaclarke commented:
@stuartmorgan okay, done, added asserts, PTAL.

stuartmorgan commented on packages/pigeon/e2e_tests/test_objc/ios/.gitignore:
```diff
@@ -31,3 +31,6 @@ Runner/GeneratedPluginRegistrant.*
 !default.pbxuser
 !default.perspectivev3
+
+Runner/dartle.h
```
Add a comment as to why these are excluded?

gaaclarke replied:
Done

stuartmorgan commented on packages/pigeon/e2e_tests/test_objc/ios/Runner/MyApi.m:
```diff
@@ -6,7 +6,10 @@
 
 @implementation MyApi
-- (ACSearchReply *)search:(ACSearchRequest *)input error:(FlutterError **)error {
+- (void)initialize:(FlutterError *_Nullable *_Nonnull)error {
```
Why is an empty `+initialize` being generated now?

gaaclarke replied:
I didn't add the parameter name for functions that have zero arguments.  This is declared as `void initialize()`.  Since no parameter name is given one isn't inserted in the selector.  `initialize:` versus `initializeError:` or `initializeCompletion:`

stuartmorgan replied:
This question was PEBCAK; I saw the `+` at the start of the line` and the word `initialize`, and somehow ignored everything else and thought this was generating an `[NSObject +initialize]` for no apparent reason.

Now that I understand what I'm looking at though: for zero-length we correctly generate these as `...WithError:` and `...WithCompletion:`. Is there a reason not to do that? I could see "consistency, because we don't do `With` in other cases", except that this is already inconsistent with other cases, so why no be inconsistent and right rather than inconsistent and wrong? (And I would argue that `...WithError:` is actually significantly more consistent with `...:error:` than having no name for error in the zero-argument case.)

gaaclarke replied:
`..WithError:` is more idiomatic.  I think we could do that with a tiny change to `_getSelectorComponents`.  I'm on board as long  as I don't have to replumb everything.  Do you think it should be like that?  I think we are running into the question again about consistency between platforms and more idiomatic objc code.  Coupled with the fact that consistency has usually meant less complex generation code.

gaaclarke replied:
I implemented `...WithError:`, it wasn't so bad.  There is some cleanup that's probably possible to make it better but it wasn't apparent to me right now.  PTAL.

stuartmorgan commented on packages/pigeon/e2e_tests/test_objc/ios/Runner/MyApi.m:
```diff
@@ -6,7 +6,10 @@
+- (void)initialize:(FlutterError *_Nullable *_Nonnull)error {
+}
+
+- (ACSearchReply *)searchRequest:(ACSearchRequest *)input error:(FlutterError **)error {
```
Is the fact that the first argument is always called "input" legacy from single-arity? We should fix that as part of the breaking change. (Sorry, I didn't notice this before because I was focused on the signature itself.)

stuartmorgan replied:
I just realized that in the test code I was looking at at the end of the review, the first argument was correctly named. Is this manually written code that's following the (previous) conventions of the generator?

gaaclarke replied:
Yea, the objc compiler doesn't enforce that argument names match between the header and the implementation.  In the generated header the argument name matches the dart declaration.  You're probably right that this was written before I started using the argument names in the generator.

stuartmorgan commented on packages/pigeon/lib/objc_generator.dart:
```diff
@@ -232,52 +234,91 @@ NSObject<FlutterMessageCodec>* ${_getCodecGetterName(options.prefix, api.name)}(
+    (String component, String argType, String argName) =>
+        '$component:($argType)$argName',
+  ).join(' ');
+  return '-($returnType)$argSignature';
```
We should follow Google style; space between `-` and `(`.

gaaclarke replied:
Done

stuartmorgan commented on packages/pigeon/lib/objc_generator.dart:
```diff
@@ -384,11 +425,22 @@ String _dictValue(
   final String apiName = _className(options.prefix, api.name);
   indent.write(
-      'void ${apiName}Setup(id<FlutterBinaryMessenger> binaryMessenger, id<$apiName> api) ');
+      'void ${apiName}Setup(id<FlutterBinaryMessenger> binaryMessenger, NSObject<$apiName>* api) ');
```
Nit: C function names should generally be verbs, and the verb is "set up" (and thus `SetUp(...)`)

gaaclarke replied:
Yea, good point.  It is consistent across platforms and unfortunately the `set` verb has special connotations in some languages too.  Like we aren't setting `up`.

stuartmorgan replied:
> It is consistent across platforms

The comment is actually applicable across platforms/languages as well, I just flagged it in the one place. (Java style guide: "Method names are typically verbs or verb phrases." Effective Dart: "PREFER starting function or method comments with third-person verbs.")

> and unfortunately the set verb has special connotations in some languages too. Like we aren't setting up.

There are a few options to avoid that:
- Stick with setup as a noun, but add a verb: `PerformSetup`/`DoSetup`/etc.
- Make the name less generic: `SetUpPigeonMethodChannels` or the like.

In this case I missed the `${apiName}` being the start, so it should be the first. For Dart it could be either.

stuartmorgan commented on packages/pigeon/test/objc_generator_test.dart:
```diff
@@ -1281,4 +1285,126 @@ void main() {
+      expect(
+          code,
+          contains(
+              '-(void)addX:(NSNumber *)arg_x y:(NSNumber *)arg_y completion:(void(^)(NSNumber*, NSError* _Nullable))completion {'));
```
Your generated code is inconsistent about `Type*` vs `Type *`. Either is fine (the latter is more common in ObjC code in general), but it should be consistent at the file level per style guide (and certainly per signature).

gaaclarke replied:
Done. Tried to move everything to `Type *`.  Should be consistent now.

stuartmorgan reviewed (changes requested):
A few small things I missed previously.

gaaclarke commented:
I think everything is addressed.  The "request review" button isn't available so just pinging on here to be safe.

rakudrama commented on packages/pigeon/lib/functional.dart:
```diff
@@ -0,0 +1,70 @@
+
+/// Adds [value] to the end of [ts].
+Iterable<T> followedByOne<T>(Iterable<T> ts, T value) sync* {
+  for (final T item in ts) {
```
`yield* ts;` is the same as this loop, but can be more efficient if `ts` is also a `sync*` iterable.
